torch/distributed/_spmd/api.py

```diff
@@ -51,3 +76,209 @@ def forward(self, *args: Tuple[object], **kwargs: Dict[str, object]) -> object:
+                if isinstance(arg, nn.Module):
+                    assert mod is None, "Only support single nn.Module for now"
+                    mod = arg
+                if isinstance(arg, torch.optim.Optimizer):
+                    assert opt is None, "Only support single Optimizer for now"
+                    opt = arg
+
+            assert mod is not None and opt is not None
+
+            # 2. Override target submodules (e.g., MoE) with dummy replacements
+            if override_map:
+                accessor = NamedMemberAccessor(mod)
+
+                for typ, override in override_map.items():
+                    for name, submodule in mod.named_modules():
+                        if isinstance(submodule, typ):
+                            accessor.swap_submodule(
+                                name, override.replacement(submodule)
+                            )
+
+            # 3. Trace statelss version of the train_step
+            params_and_buffers = {
+                **dict(mod.named_parameters(remove_duplicate=False)),
+                **dict(mod.named_buffers(remove_duplicate=False)),
+            }
+
+            opt_states, spec = pytree.tree_flatten(dict(opt.state))
+            named_states = {}
+            for n, p in params_and_buffers.items():
+                if p in opt.state:
+                    named_states[n] = opt.state[p]
+
+            if dist.get_rank() == 0:
+                print(f"====== {len(opt.state)}, {opt.state}")
+
+            def stateless_func(
+                func, args, kwargs, named_states, params_and_buffers
+            ):
+                with stateless._reparametrize_module(
+                    mod, params_and_buffers
+                ), _rematerialize_optimizer(
+                    opt, named_states, params_and_buffers
+                ):
+                    func(*args, **kwargs)
+                    return list(mod.parameters())
+
+            # parameters are at the end of placeholders
+            # FIXME(@mrshenli): let make_fx ignore nn.Module in args? or it should
+            # eliminate unused placeholders?
+            # FIXME: Using symbolic tracing to work around. Otherwise it hits
+            # shape mismatch error, as we use local inputs to trace local graph
+            # and use DTensor to expand operators, where DTensor's shape is the
+            # global shape.
+            with _enable_compile():
+                gm = make_fx(
+                    partial(stateless_func, func),
+                    tracing_mode="symbolic",
+                    _allow_non_fake_inputs=True,
+                )(args, kwargs, named_states, params_and_buffers)
+
+            # 4. Use DTensor to insert collectives
+            gm, name_to_spec = _dtensor_expand(
+                gm, args, kwargs, named_states, params_and_buffers
+            )
+
+            # FIXME(@mrshenli): debug only
+            if dist.get_rank() == 0:
+                gm.graph.print_tabular()
```

Comment:
adding optimizer into the graph does not seem to trigger DTensor partial->replicate allreduce. Will get back to this later. 

```
opcode         name          target                   args                             kwargs                          
-------------  ------------  -----------------------  -------------------------------  ---------------------------------------------------------------------------
---------------------------------------------------------------------------                                                                                       
placeholder    arg0_1        arg0_1                   ()                               {}
placeholder    arg0_2        arg0_2                   ()                               {}
placeholder    arg0_3        arg0_3                   ()                               {}
placeholder    arg2_1        arg2_1                   ()                               {}
placeholder    arg2_2        arg2_2                   ()                               {}
placeholder    arg2_3        arg2_3                   ()                               {}
placeholder    arg2_4        arg2_4                   ()                               {}
placeholder    arg2_5        arg2_5                   ()                               {}
placeholder    arg2_6        arg2_6                   ()                               {}
placeholder    arg3_1        arg3_1                   ()                               {}
placeholder    arg3_2        arg3_2                   ()                               {}
call_function  t_4           aten.t.default           (arg3_1,)                        {}
call_function  addmm_1       aten.addmm.default       (arg3_2, arg0_3, t_4)            {}
call_function  sum_3         aten.sum.default         (addmm_1,)                       {}
call_function  ones_like_1   aten.ones_like.default   (sum_3,)                         {'dtype': torch.float32, 'layout': torch.strided, 'device': device(type='cu
da', index=0), 'pin_memory': False, 'memory_format': torch.preserve_format}                                                                                       
call_function  expand_1      aten.expand.default      (ones_like_1, [4, 10])           {}
call_function  t_5           aten.t.default           (expand_1,)                      {}                              
call_function  slice_1       aten.slice.Tensor        (t_5, 1, 0, 2)                   {}
call_function  clone         aten.clone.default       (slice_1,)                       {}
call_function  mm_1          aten.mm.default          (clone, arg0_3)                  {}
call_function  t_6           aten.t.default           (mm_1,)                          {}
call_function  sum_4         aten.sum.dim_IntList     (expand_1, [0], True)            {}
call_function  view_1        aten.view.default        (sum_4, [10])                    {}
call_function  detach_4      aten.detach.default      (view_1,)                        {}
call_function  detach_5      aten.detach.default      (detach_4,)                      {}
call_function  detach_6      aten.detach.default      (detach_5,)                      {}
call_function  detach_7      aten.detach.default      (detach_6,)                      {}
call_function  t_7           aten.t.default           (t_6,)                           {}
call_function  detach_8      aten.detach.default      (t_7,)                           {}
call_function  detach_9      aten.detach.default      (detach_8,)                      {}
call_function  detach_10     aten.detach.default      (detach_9,)                      {}
call_function  detach_11     aten.detach.default      (detach_10,)                     {}
call_function  add_          aten.add_.Tensor         (arg2_1, 1)                      {}
call_function  mul_          aten.mul_.Tensor         (arg2_2, 0.9)                    {}
call_function  add__1        aten.add_.Tensor         (mul_, detach_11)                {'alpha': 0.09999999999999998}
call_function  mul__1        aten.mul_.Tensor         (arg2_3, 0.999)                  {}
call_function  addcmul_      aten.addcmul_.default    (mul__1, detach_11, detach_11)   {'value': 0.0010000000000000009}
call_function  pow_5         aten.pow.Scalar          (0.9, add_)                      {}
call_function  rsub_4        aten.rsub.Scalar         (pow_5, 1)                       {}
call_function  pow_6         aten.pow.Scalar          (0.999, add_)                    {}
call_function  rsub_5        aten.rsub.Scalar         (pow_6, 1)                       {}
call_function  reciprocal_4  aten.reciprocal.default  (rsub_4,)                        {}
call_function  mul_6         aten.mul.Tensor          (reciprocal_4, 0.01)             {}
call_function  neg_2         aten.neg.default         (mul_6,)                         {}
call_function  sqrt_4        aten.sqrt.default        (rsub_5,)                        {}
call_function  sqrt_5        aten.sqrt.default        (addcmul_,)                      {}
call_function  mul_7         aten.mul.Tensor          (sqrt_4, neg_2)                  {}
call_function  div_2         aten.div.Tensor          (sqrt_5, mul_7)                  {}
call_function  reciprocal_5  aten.reciprocal.default  (neg_2,)                         {}
call_function  mul_8         aten.mul.Tensor          (reciprocal_5, 1e-08)            {}
call_function  add__2        aten.add_.Tensor         (div_2, mul_8)                   {}
call_function  addcdiv_      aten.addcdiv_.default    (arg3_1, add__1, add__2)         {}
call_function  add__3        aten.add_.Tensor         (arg2_4, 1)                      {}
call_function  mul__2        aten.mul_.Tensor         (arg2_5, 0.9)                    {}
call_function  add__4        aten.add_.Tensor         (mul__2, detach_7)               {'alpha': 0.09999999999999998}
call_function  mul__3        aten.mul_.Tensor         (arg2_6, 0.999)                  {}
call_function  addcmul__1    aten.addcmul_.default    (mul__3, detach_7, detach_7)     {'value': 0.0010000000000000009}
call_function  pow_7         aten.pow.Scalar          (0.9, add__3)                    {}
call_function  rsub_6        aten.rsub.Scalar         (pow_7, 1)                       {}
call_function  pow_8         aten.pow.Scalar          (0.999, add__3)                  {}
call_function  rsub_7        aten.rsub.Scalar         (pow_8, 1)                       {}
call_function  reciprocal_6  aten.reciprocal.default  (rsub_6,)                        {}
call_function  mul_9         aten.mul.Tensor          (reciprocal_6, 0.01)             {}
call_function  neg_3         aten.neg.default         (mul_9,)                         {}
call_function  sqrt_6        aten.sqrt.default        (rsub_7,)                        {}
call_function  sqrt_7        aten.sqrt.default        (addcmul__1,)                    {}
call_function  mul_10        aten.mul.Tensor          (sqrt_6, neg_3)                  {}
call_function  div_3         aten.div.Tensor          (sqrt_7, mul_10)                 {}
call_function  reciprocal_7  aten.reciprocal.default  (neg_3,)                         {}
call_function  mul_11        aten.mul.Tensor          (reciprocal_7, 1e-08)            {}
call_function  add__5        aten.add_.Tensor         (div_3, mul_11)                  {}
call_function  addcdiv__1    aten.addcdiv_.default    (arg3_2, add__4, add__5)         {}
output         output        output                   ([None, addcdiv_, addcdiv__1],)  {}
```

Reply:
update: when using SGD, allreduce can be correctly triggered, but not Adam